gax/src/main/java/com/google/api/gax/retrying/IgnoreRetryingContextResultRetryAlgorithm.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import java.util.concurrent.CancellationException;
+
+/**
+ * Default implementation of {@link ResultRetryAlgorithmWithContext} that ignores the {@link
+ * RetryingContext}. This is used to wrap {@link ResultRetryAlgorithm} instances to create a {@link
+ * ResultRetryAlgorithmWithContext} when one is required.
+ */
+class IgnoreRetryingContextResultRetryAlgorithm<ResponseT>
```

Comment:
Why are these `Ignore*` algorithms needed? Can we just add the new functionality (of context-awareness) to the existing classes (`ExponentialRetryAlgorithm`, `BasicResultRetryAlgorithm`)? Just make them ignore the context part if context is `null`, or make it a config parameter (irnore or not ignore context) in the classes constructor. 

The idea is to not create new classes if they are not necessary. Java has reputation of being a very verbose langauge, this is one of the examples where we can avoid unnecessary verbosity.

Reply:
The reason that we need these classes is the public constructor [here](https://github.com/olavloite/gax-java/blob/f3225490740230272f073c83a83a95c690e3bb12/gax/src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java#L62):

```java
  public RetryAlgorithm(
      ResultRetryAlgorithm<ResponseT> resultAlgorithm, TimedRetryAlgorithm timedAlgorithm) {
    this.resultAlgorithm =
        new IgnoreRetryingContextResultRetryAlgorithm<>(checkNotNull(resultAlgorithm));
    this.timedAlgorithm =
        new IgnoreRetryingContextTimedRetryAlgorithm(checkNotNull(timedAlgorithm));
  }
```

* Both `ResultRetryAlgorithm` and `TimedRetryAlgorithm` are interfaces, so we can't add any new methods to those without introducing a breaking change.
* Removing or changing the public constructor to one that only takes `ContextAware` algorithms is also a breaking change.